README.md
# NAME

MooseX::Role::Flyweight - Automatically memoize your Moose objects for reuse

# VERSION

version 1.03

# SYNOPSIS

Compose MooseX::Role::Flyweight into your Moose class.

    package Glyph::Character;
    use Moose;
    with 'MooseX::Role::Flyweight';

    has 'c' => (is => 'ro', required => 1);

    sub draw {
        my ($self, $context) = @_;
        ...
    }

    # Optional: override normalizer()
    sub normalizer {
        my ($class, $init_args) = @_;
        return $init_args->{c};
    }

Get cached object instances by calling `instance()` instead of `new()`.

    # the same initialisation arguments produces the same object
    $shared_object   = Glyph::Character->instance( %init_args );
    $same_object     = Glyph::Character->instance( %init_args );

    # different initialisation arguments produces a different object
    $another_object  = Glyph::Character->instance( %diff_args );

    # new() still works but its objects are not shared
    $unshared_object = Glyph::Character->new( %init_args );

# DESCRIPTION

_A million tiny objects can weigh a ton._

Instead of creating a multitude of identical copies of objects, a flyweight
is a memoized instance that may be reused in multiple contexts simultaneously
to minimize memory usage. And due to the cost of constructing objects the
reuse of flyweights has the potential to speed up your code.

MooseX::Role::Flyweight is a Moose role that enables your Moose class to
automatically manage a cache of reusable instances. In other words, the class
becomes its own flyweight factory.

## Flyweight v. Singleton

MooseX::Role::Flyweight provides an `instance()` method which looks similar
to [MooseX::Singleton](https://metacpan.org/module/MooseX::Singleton). This is in part because MooseX::Role::Flyweight
departs from the original "Gang of Four" design pattern in that the role of
the Flyweight Factory has been merged into the Flyweight class itself. But the
choice of the method name was based on MooseX::Singleton.

While MooseX::Role::Flyweight and MooseX::Singleton look similar, understanding
their intentions will highlight their differences:

- Singleton

    MooseX::Singleton limits the number of instances allowed for that class to
    ONE. For this reason, its `instance()` method does not accept construction
    arguments and will always return the same instance. If arguments are required
    for construction, then you will need to call its `initialize()` method.

- Flyweight

    MooseX::Role::Flyweight is used to facilitate the reuse of objects to reduce
    the cost of having many instances. The number of instances created will be
    reduced, but it does not set a limit on how many instances are allowed. Its
    `instance()` method does accept construction arguments because it is
    responsible for managing the construction of new instances when it finds
    that it cannot reuse an existing one.

# METHODS

## instance

    $instance = My::Flyweight->instance( %init_args );

This class method returns an instance that has been constructed from the given
arguments. The first time it is called with a given set of arguments it will
construct the object and cache it. On subsequent calls with the equivalent set
of arguments it will reuse the existing object by retrieving it from the cache.

The arguments may be in any form that `new()` will accept. This is normally a
hash or hash reference of named parameters. Non-hash(ref) arguments are also
possible if you have defined your own `BUILDARGS` class method to handle them
(see [Moose::Manual::Construction](https://metacpan.org/module/Moose::Manual::Construction)).

Note that instances that are constructed by calling `new()` directly do not
get cached and therefore will never be returned by this method.

## normalizer

    $instance_identifier_string = My::Flyweight->normalizer( $init_args_hashref );

This class method generates the keys used by `instance()` to identify objects
for storage and retrieval in the cache. It is passed the arguments used for
construction as a hashref (after they have passed through `BUILDARGS`). It
returns a string representation of those arguments as the key. Equivalent
arguments should result in the same string.

Note that this does not handle blessed references as arguments.

Generally you should not need to access this method directly. The only reason
you would want to know about this method is if you want to change the way it
generates the cache keys, in which case you should wrap or override this
method in your class that consumes this role.

# NOTES ON USAGE

## Flyweights should be immutable

Your flyweight object attributes should be read-only. It is dangerous to have
mutable flyweight objects because it means you may get something you don't
expect when you retrieve it from the cache the next time.

    my $flight = Flight->instance( destination => 'Australia' );
    $flight->set_destination('Antarctica');

    # ... later, in another context
    my $flight = Flight->instance( destination => 'Australia' );
    die 'hypothermia'  if $flight->destination eq 'Antarctica';

Value objects are the type of objects that are suited as flyweights.

## Argument normalization

Instances are identified for reuse based on the equivalency of the named
parameters used for construction as interpreted by `normalizer()`.

Factors to consider when determining equivalency:

- There is no distinction between hash and hashref (and non-hash) arguments.

        # same object is returned
        $obj1 = My::Flyweight->instance( attr => 'value' );
        $obj2 = My::Flyweight->instance({attr => 'value'});
- The order of named parameters does not affect equivalency.

    The keys in the hash(ref) are sorted, which means that the same string will
    always be produced for the same named parameters regardless of the order
    they are given.

        # same object is returned
        $obj1 = My::Flyweight->instance( attr1 => 1, attr2 => 2 );
        $obj2 = My::Flyweight->instance( attr2 => 2, attr1 => 1 );

On the other hand, `normalizer()` does not handle:

- Unused construction parameters.

    You can use [MooseX::StrictConstructor](https://metacpan.org/module/MooseX::StrictConstructor) to prevent this.

        # different objects with same values returned
        $obj1 = My::Flyweight->instance( attr => 'value' );
        $obj2 = My::Flyweight->instance( attr => 'value', unused_attr => 'value' );

- Default attribute values.

    You can extend/override `normalizer()` to handle this if you wish.

        # different objects with same values returned
        $obj1 = My::Flyweight->instance( attr1 => 'value' );
        $obj2 = My::Flyweight->instance( attr1 => 'value', attr2 => 'default' );

## Garbage collection of cached objects

The cache uses weak references to the objects so that the cache references
do not prevent the objects from being garbage collected. This means that an
object in the cache will be destroyed when all other references to it go out
of scope.

    my $obj = My::Flyweight->instance(%args);
    # $obj is in the cache
    undef $obj;
    # $obj is garbage collected and disappears from the cache

# AUTHOR

Steven Lee <stevenwh.lee@gmail.com>

# ACKNOWLEDGEMENTS

Thanks to Mark Stosberg (MARKSTOS) for suggesting to explain the difference
between MooseX::Role::Flyweight and MooseX::Singleton.

# SEE ALSO

[Perl Design Patterns](http://www.perl.com/pub/2003/06/13/design1.html)

[Memoize](https://metacpan.org/module/Memoize)

# COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Steven Lee.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for MouseX-Role-Flyweight

1.04    2014-11-05
        Fork from MooseX::Role::Flyweight.

1.03    2014-02-14
        Use a simple hash instead of MooseX::ClassAttribute.
        Apply perltidy and perlcritic.

1.02    2013-10-09
        Change normalizer() to accept post-BUILARGS hashrefs so users don't
        need to remember to do it. But it is still backwards compatible.

1.01    2013-09-22
        Upgrade MooseX::ClassAttribute version prereq to 0.27 to work with
        Moose 2.08+

1.00    2013-09-17
        Promote to stable release.

0.007   2013-08-22
        Documentation adjustments

0.006   2013-05-17
        Fix tests with insecure dependency in taint mode

0.005   2013-05-10
        Weaken cache references

0.004   2013-04-03
        Add documentation to highlight differences with MooseX::Singleton
        and notes on usage

0.003   2013-04-02
        Minor bug fixes
        Add JSON::XS prerequisite recommendation

0.002   2013-04-01
        Fixed distribution configuration and set $VERSION

0.001   2013-03-29
        Initial release


t/lib/Flyweight/Test1.pm
package Flyweight::Test1;
use Moose;
with 'MooseX::Role::Flyweight';

has id    => ( is => 'ro', isa => 'Int', default => 0 );
has value => ( is => 'ro', isa => 'Str', default => '' );

around BUILDARGS => sub {
    my $orig  = shift;
    my $class = shift;
    my @args  = @_ == 1 && !ref( $_[0] ) ? ( id => $_[0] ) : @_;

    return $class->$orig(@args);
};

__PACKAGE__->meta->make_immutable;
1;


t/lib/Flyweight/Test2.pm
package Flyweight::Test2;
use Moose;
with 'MooseX::Role::Flyweight';

has id => ( is => 'ro', isa => 'Int', required => 1 );

# constructor can accept id as a single value
around BUILDARGS => sub {
    my $orig  = shift;
    my $class = shift;
    return @_ == 1 && !ref $_[0]
        ? $class->$orig( id => $_[0] )
        : $class->$orig(@_);
};

__PACKAGE__->meta->make_immutable;
1;


t/lib/Flyweight/Test3.pm
package Flyweight::Test3;
use Moose;
with 'MooseX::Role::Flyweight';

has id => (
    is       => 'ro',
    isa      => 'Int',
    required => 1,
);

has attr => (
    is      => 'ro',
    default => 1,
);

has _init_attr => (
    is       => 'ro',
    init_arg => 'init_attr',
    default  => 1,
);

has _uninit_attr => (
    is       => 'ro',
    init_arg => undef,
    default  => 1,
);

has _private_attr => (
    is      => 'rw',
    default => 1,
);

has _lazy_attr => (
    is      => 'ro',
    lazy    => 1,
    default => 1,
);

around normalizer => sub {
    my ( $orig, $class, $args ) = @_;

    # handle invalid attributes
    my %attributes = map { $_->init_arg => undef }
        grep { defined $_->init_arg } $class->meta->get_all_attributes;

    my @unknown = grep { !exists $attributes{$_} } keys %$args;
    confess "Found unknown attribute(s): @unknown"
        if @unknown > 0;

    # add default attribute values
    foreach my $attr ( $class->meta->get_all_attributes ) {
        $args->{ $attr->init_arg } = $attr->default
            if (
               $attr->has_init_arg
            && $attr->has_default
            && !defined $args->{ $attr->init_arg }  # arg is already set
            && $attr->init_arg !~ /^\_/             # marked as private
            && !$attr->is_lazy                      # lazy needs obj to be built
            && !$attr->is_default_a_coderef
            );
    }

    return $class->$orig($args);
};

__PACKAGE__->meta->make_immutable;
1;


lib/MooseX/Role/Flyweight.pm
package MooseX::Role::Flyweight;
# ABSTRACT: Automatically memoize your Moose objects for reuse

=for :badge
=for html
<a href="https://travis-ci.org/stevenl/MooseX-Role-Flyweight"><img src="https://travis-ci.org/stevenl/MooseX-Role-Flyweight.svg?branch=master" alt="Build Status"></a>
<a href='https://coveralls.io/r/stevenl/MooseX-Role-Flyweight?branch=master'><img src='https://coveralls.io/repos/stevenl/MooseX-Role-Flyweight/badge.png?branch=master' alt='Coverage Status' /></a>

=head1 SYNOPSIS

Compose MooseX::Role::Flyweight into your Moose class.

    package Glyph::Character;
    use Moose;
    with 'MooseX::Role::Flyweight';

    has 'c' => ( is => 'ro', required => 1 );

    sub draw {
        my ( $self, $context ) = @_;
        ...
    }

    # Optional: override normalizer()
    sub normalizer {
        my ( $class, $init_args ) = @_;
        return $init_args->{c};
    }

Get cached object instances by calling C<instance()> instead of C<new()>.

    # the same initialisation arguments produces the same object
    $shared_object   = Glyph::Character->instance(%init_args);
    $same_object     = Glyph::Character->instance(%init_args);

    # different initialisation arguments produces a different object
    $another_object  = Glyph::Character->instance(%diff_args);

    # new() still works but its objects are not shared
    $unshared_object = Glyph::Character->new(%init_args);

=head1 DESCRIPTION

I<A million tiny objects can weigh a ton.>

Instead of creating a multitude of identical copies of objects, a flyweight
is a memoized instance that may be reused in multiple contexts simultaneously
to minimize memory usage. And due to the cost of constructing objects the
reuse of flyweights has the potential to speed up your code.

MooseX::Role::Flyweight is a Moose role that enables your Moose class to
automatically manage a cache of reusable instances. In other words, the class
becomes its own flyweight factory.

=head2 Flyweight v. Singleton

MooseX::Role::Flyweight provides an C<instance()> method which looks similar
to L<MooseX::Singleton>. This is in part because MooseX::Role::Flyweight
departs from the original "Gang of Four" design pattern in that the role of
the Flyweight Factory has been merged into the Flyweight class itself. But the
choice of the method name was based on MooseX::Singleton.

While MooseX::Role::Flyweight and MooseX::Singleton look similar, understanding
their intentions will highlight their differences:

=begin :list

= Singleton

MooseX::Singleton limits the number of instances allowed for that class to
ONE. For this reason, its C<instance()> method does not accept construction
arguments and will always return the same instance. If arguments are required
for construction, then you will need to call its C<initialize()> method.

= Flyweight

MooseX::Role::Flyweight is used to facilitate the reuse of objects to reduce
the cost of having many instances. The number of instances created will be
reduced, but it does not set a limit on how many instances are allowed. Its
C<instance()> method does accept construction arguments because it is
responsible for managing the construction of new instances when it finds
that it cannot reuse an existing one.

=end :list

=cut

use 5.006;
use JSON 2.00 ();  # works with JSON::XS
use Moose::Role;
use namespace::autoclean;
use Scalar::Util ();

my $JSON = JSON->new->utf8->canonical;

our %INSTANCES;

=method instance

    $instance = My::Flyweight->instance(%init_args);

This class method returns an instance that has been constructed from the given
arguments. The first time it is called with a given set of arguments it will
construct the object and cache it. On subsequent calls with the equivalent set
of arguments it will reuse the existing object by retrieving it from the cache.

The arguments may be in any form that C<new()> will accept. This is normally a
hash or hash reference of named parameters. Non-hash(ref) arguments are also
possible if you have defined your own C<BUILDARGS> class method to handle them
(see L<Moose::Manual::Construction>).

Note that instances that are constructed by calling C<new()> directly do not
get cached and therefore will never be returned by this method.

=cut

sub instance {
    my ( $class, @args ) = @_;
    my $args = $class->BUILDARGS(@args);
    my $key  = $class->normalizer($args);

    # return the existing instance
    return $INSTANCES{$class}{$key}
        if defined $INSTANCES{$class}{$key};

    # create a new instance
    my $instance = $class->new(@args);
    $INSTANCES{$class}{$key} = $instance;
    Scalar::Util::weaken $INSTANCES{$class}{$key};

    return $instance;
}

=method normalizer

    $instance_identifier_string = My::Flyweight->normalizer($init_args_hashref);

This class method generates the keys used by C<instance()> to identify objects
for storage and retrieval in the cache. It is passed the arguments used for
construction as a hashref (after they have passed through C<BUILDARGS>). It
returns a string representation of those arguments as the key. Equivalent
arguments should result in the same string.

Note that this does not handle blessed references as arguments.

Generally you should not need to access this method directly. The only reason
you would want to know about this method is if you want to change the way it
generates the cache keys, in which case you should wrap or override this
method in your class that consumes this role.

=cut

sub normalizer {
    my ( $class, @args ) = @_;
    my $args =
        ( @args > 1 || ref( $args[0] ) ne 'HASH' )
        ? $class->BUILDARGS(@args)
        : $args[0];

    return $JSON->encode($args);
}

1;

=head1 NOTES ON USAGE

=head2 Flyweights should be immutable

Your flyweight object attributes should be read-only. It is dangerous to have
mutable flyweight objects because it means you may get something you don't
expect when you retrieve it from the cache the next time.

    my $flight = Flight->instance( destination => 'Australia' );
    $flight->set_destination('Antarctica');

    # ... later, in another context
    my $flight = Flight->instance( destination => 'Australia' );
    die 'hypothermia'  if $flight->destination eq 'Antarctica';

Value objects are the type of objects that are suited as flyweights.

=head2 Argument normalization

Instances are identified for reuse based on the equivalency of the named
parameters used for construction as interpreted by C<normalizer()>.

Factors to consider when determining equivalency:

=begin :list

* There is no distinction between hash and hashref (and non-hash) arguments.

    # same object is returned
    $obj1 = My::Flyweight->instance( attr => 'value' );
    $obj2 = My::Flyweight->instance( { attr => 'value' } );

* The order of named parameters does not affect equivalency.

The keys in the hash(ref) are sorted, which means that the same string will
always be produced for the same named parameters regardless of the order
they are given.

    # same object is returned
    $obj1 = My::Flyweight->instance( attr1 => 1, attr2 => 2 );
    $obj2 = My::Flyweight->instance( attr2 => 2, attr1 => 1 );

=end :list

On the other hand, C<normalizer()> does not handle:

=begin :list

* Unused construction parameters.

You can use L<MooseX::StrictConstructor> to prevent this.

    # different objects with same values returned
    $obj1 = My::Flyweight->instance( attr => 'value' );
    $obj2 = My::Flyweight->instance( attr => 'value', unused_attr => 'value' );

* Default attribute values.

You can extend/override C<normalizer()> to handle this if you wish.

    # different objects with same values returned
    $obj1 = My::Flyweight->instance( attr1 => 'value' );
    $obj2 = My::Flyweight->instance( attr1 => 'value', attr2 => 'default' );

=end :list

=head2 Garbage collection of cached objects

The cache uses weak references to the objects so that the cache references
do not prevent the objects from being garbage collected. This means that an
object in the cache will be destroyed when all other references to it go out
of scope.

    my $obj = My::Flyweight->instance(%args);
    # $obj is in the cache
    undef $obj;
    # $obj is garbage collected and disappears from the cache

=head1 ACKNOWLEDGEMENTS

Thanks to Mark Stosberg (MARKSTOS) for suggesting to explain the difference
between MooseX::Role::Flyweight and MooseX::Singleton.

=head1 SEE ALSO

L<Perl Design Patterns|http://www.perl.com/pub/2003/06/13/design1.html>

L<Memoize>

=cut
